File: ccc13s4.js
function solve(comparisons, height, target) {
  let queue = [];
  queue.push(height);
  while(queue.length > 0) {
    height = queue.shift();
    if (height === target) {
      return true;
    }
    comparisons[height].forEach(shorter => queue.push(shorter));
  }
}


const gets = () => require('readline-sync').question();
const print = (x) => console.log(x);

const [numS, numCompared] = gets().split(' ').map(n => Number(n));

let x, y;
let comparisons = {};
for (let i = 0; i < numCompared; i++) {
  [x, y] = gets().split(' ').map(n => Number(n));
  if (comparisons[x]) comparisons[x].push(y);
  else comparisons[x] = [y];
  if (!comparisons[y]) comparisons[y] = [];
}

const [p, q] = gets().split(' ').map(n => Number(n));
if (solve(comparisons, p, q)) print('yes');
else if (solve(comparisons, q, p)) print('no');
else print('unknown');

File: ccc13s4.cpp
#include <bits/stdc++.h>

using namespace std;
typedef vector<int> vi;

vi comparisons[1000001]{};

bool isTaller(int height, int target) {
  queue <int> q;
  q.push(height);
  while (!q.empty()) {
    height = q.front();
    q.pop();
    if (height == target) {
      return true;
    }
    for (int i = 0; i < comparisons[height].size(); i++) {
      q.push(comparisons[height][i]);
    }
  }
  return false;
}


int main() {
  ios_base::sync_with_stdio(false);
  cin.tie(NULL);
  
  int numStudents, numCompared;
  int x, y;

  cin >> numStudents >> numCompared;

  for (int i = 0; i < numCompared; i++) {
    cin >> x >> y;
    comparisons[x].push_back(y);
  }

  cin >> x >> y;
  if (isTaller(x, y)) cout << "yes" << '\n';
  else if (isTaller(y, x)) cout << "no" << '\n';
  else cout << "unknown" << '\n';

  return 0;
}